module/VuFind/src/VuFind/Content/Covers/Google.php

```diff
@@ -86,23 +73,44 @@ public function getUrl($key, $size, $ids)
         // Construct the request URL and make the HTTP request:
         $url = 'https://books.google.com/books?jscmd=viewapi&' .
                'bibkeys=ISBN:' . $ids['isbn']->get13() . '&callback=addTheCover';
-        $result = $this->getHttpClient($url)->send();
 
-        // If the request was successful and we can extract a valid response...
-        if ($result->isSuccess()
-            && preg_match('/^[^{]*({.*})[^}]*$/', $result->getBody(), $matches)
-        ) {
-            // convert \x26 or \u0026 to &
+        $decodeCallback = function (\Laminas\Http\Response $response, $url) {
+            if (!preg_match(
+                '/^[^{]*({.*})[^}]*$/',
+                $response->getBody(),
+                $matches
+            )
+            ) {
+                throw new HttpDownloadException(
+                    'Invalid response body (raw)',
+                    $url,
+                    $response->getStatusCode(),
+                    $response->getHeaders(),
+                    $response->getBody()
+                );
+            }
+
             $json = json_decode(
                 str_replace(['\\x26', '\\u0026'], '&', $matches[1]),
                 true
             );
 
-            // find the first thumbnail URL and process it:
-            foreach ((array)$json as $current) {
-                if (isset($current['thumbnail_url'])) {
-                    return $current['thumbnail_url'];
-                }
+            if ($json === null) {
+                throw new HttpDownloadException(
+                    'Invalid response body (json)',
+                    $url,
+                    $response->getStatusCode(),
+                    $response->getHeaders(),
+                    $response->getBody()
+                );
+            }
+        };
+
+        $json = $this->cachingDownloader->download($url, [], $decodeCallback);
```

Comment:
We shouldn't assume that `cachingDownloader` is set here, since it is set optionally. We should either throw an exception or have a non-cached fallback.

Reply:
I still think a simple safety check would be a good idea, since we can't guarantee the code will be built by our factory. (For example, if somebody has built a custom factory in an older version and then upgrades, we don't want things to mysteriously break -- not that I think this is particularly likely in this case, but that's the principle behind my suggestion :-) ).